### PR TITLE
feat: Add difference and intersect  BigQuery classes

### LINKED ISSUE #87: Except and Intersect
Similar to `UNION`, the intersect and except methods need to be overwritten.

We would need to replace `INTERSECT` with `INTERSECT DISTINCT`.

I don't think I fully understand the different implementations for Union and Intersect? Would I need to recreate the operations and overwrite the whole `to_ast` method?

Edit: I think the last point is clear now. I was looking at different versions quite tricky to compare which ibis version is being patched where :D

### PR DESCRIPTION
**Pull request by saschahofmann**
feat: Should change the default compile SQL from `INTERSECT` to `INTERSECT DISTINCT` and same for `EXCEPT`

Closes #87

### CODE REVIEW

**Comment by saschahofmann** [comment]
I still don't know what would be a easy way to play around with this locally? Do you have a notebook or so you're using when developing ?

**Comment by saschahofmann** [comment]
@tswast is there a changelog or something I should  update? Also should I add tests for this?

**Comment by tswast** [comment]
> @tswast is there a changelog or something I should update?

Release-Please should see the `feat: ` in the PR / commit and send a PR to update the changelog as part of the next release.

> Also should I add tests for this?

Yes, please. :-) I'd at least like to see unit tests in https://github.com/ibis-project/ibis-bigquery/blob/main/tests/unit/test_compiler.py that check if the expected SQL is generated when used.

**Comment by saschahofmann** [comment]
Ok I added the test. For some reason I can't run the tests locally, getting a `AttributeError: module 'sqlalchemy' has no attribute 'Binary'`.

Do we need to lock the sqlalchemy version? 

I created the ibis-bigquery-dev env and it seems to have installed:
```sqlalchemy                1.4.20           py38h96a0964_0    conda-forge```

**Comment by saschahofmann** [comment]
Ok seems like it's still not working. If you could help me out with running the tests locally, I'll figure out why it's not using the  `difference_class` etc.

**Comment by saschahofmann** [comment]
@tswast I downgraded SQLalchemy to `1.3.23` now the previous error disappeared but I still can't run the tests. This time it fails with `AttributeError: module 'ibis.sql.compiler' has no attribute 'Intersection'` but again in CI it seems to work any idea what's different between the versions?

**Comment by tswast** [comment]
> AttributeError: module 'sqlalchemy' has no attribute 'Binary'

This is a know issue in Ibis (not really their fault, as SQLAlchemy made breaking changes in 1.4). I believe it has been fixed in https://github.com/ibis-project/ibis/pull/2690

> AttributeError: module 'ibis.sql.compiler' has no attribute 'Intersection'

Hmmm... I don't see how your change would have caused this. Ibis has been doing a ton of refactoring, so we probably need to adjust some imports and add fallbacks for older versions.

**Comment by saschahofmann** [comment]
Ok so the error probably happens independently from the version because Ibis isnt exporting the classes in https://github.com/ibis-project/ibis/blob/master/ibis/backends/base/sql/compiler/__init__.py 

I will include that in another PR there but another problem is that I can't find the Intersection class in 1.3.

**Comment by saschahofmann** [comment]
Thanks for jumping on this! I think it'd be easier to close this and create a new PR once ibis-project/ibis#4006 is done?

**Comment by tswast** [comment]
> Thanks for jumping on this! I think it'd be easier to close this and create a new PR once https://github.com/ibis-project/ibis/pull/4006 is done?

I suspect we might be able to use this PR once #4006 is done. Right now I'm getting some failed unit tests when I use my patch, so it's been helpful in development of that PR.

File: ibis_bigquery/backports.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+    _keyword = "INTERSECT"
+
+    def _get_keyword_list(self):
+        return [self._keyword] * (len(self.tables) - 1)
+
+
+# Override Compiler to use overrides for Difference and Intersections. See
+# https://github.com/ibis-project/ibis-bigquery/issues/87 and
+# https://github.com/ibis-project/ibis/issues/3863
+if IBIS_VERSION <= IBIS_3_0_2_VERSION:
```
Comment (tswast) [inline ibis_bigquery/backports.py:59]: Note to self, keep this in sync if there are any changes to https://github.com/ibis-project/ibis/pull/4006

**Pull request merged**